xe/thinCLIProtocol.cs

```diff
@@ -49,26 +53,117 @@ public class Config
         public int Port { get; set; } = 443;
         public int BlockSize { get; set; } = 65536;
         public bool Debug { get; set; }
+        public bool NoWarnNewCertificates { get; set; }
+        public bool NoWarnCertificates { get; set; }
         public List<string> EnteredParamValues { get; } = new List<string>();
     }
 
-    public static class Transport
+    public class Transport
     {
+        private readonly Config _conf;
+        private readonly object _certificateValidationLock = new object();
+
+        public Transport(Config conf)
+        {
+            _conf = conf;
+        }
+
         // The following method is invoked by the RemoteCertificateValidationDelegate.
-        private static bool ValidateServerCertificate(
+        private bool ValidateServerCertificate(
               object sender,
               X509Certificate certificate,
               X509Chain chain,
               SslPolicyErrors sslPolicyErrors)
         {
-            // Do allow this client to communicate with unauthenticated servers.
-            return true;
+            if (sslPolicyErrors == SslPolicyErrors.None)
+                return true;
+
+            lock (_certificateValidationLock)
+            {
+                var req = (SslStream)sender;
+
+                string trustEvaluation = Messages.CERTIFICATE_NOT_TRUSTED;
+                try
+                {
+                    var cert2 = new X509Certificate2(certificate);
+                    if (new X509Chain(true).Build(cert2) || cert2.Verify())
+                        trustEvaluation = Messages.CERTIFICATE_TRUSTED;
+                }
+                catch (CryptographicException)
+                {
+                    //ignore
```

Comment:
I'd print out to the user that there was no certificate when they have `NoWarnCertificates = false`